ml-agents/mlagents/trainers/ghost/trainer.py

```diff
@@ -0,0 +1,253 @@
+        :param trainer: The trainer of the policy/policies being trained with self_play
+        :param brain_name: The name of the brain associated with trainer config
+        :param reward_buff_cap: Max reward history to track in the reward buffer
+        :param trainer_parameters: The parameters for the trainer (dictionary).
+        :param training: Whether the trainer is set for training.
+        :param run_id: The identifier of the current run
+        """
+
+        super(GhostTrainer, self).__init__(
+            brain_name, trainer_parameters, training, run_id, reward_buff_cap
+        )
+
+        self.trainer = trainer
+
+        self.internal_policy_queues: List[AgentManagerQueue[Policy]] = []
+        self.learning_policy_queues: Dict[str, AgentManagerQueue[Policy]] = {}
+        self.internal_trajectory_queues: List[AgentManagerQueue[Trajectory]] = []
+
+        # assign ghost's stats collection to wrapped trainer's
+        self.stats_reporter = self.trainer.stats_reporter
+
+        self_play_parameters = trainer_parameters["self_play"]
+        self.window = self_play_parameters["window"]
+        self.play_against_current_self_ratio = self_play_parameters[
+            "play_against_current_self_ratio"
+        ]
+        self.steps_between_snapshots = self_play_parameters["snapshot_per"]
+
+        self.policies: Dict[str, TFPolicy] = {}
+        self.policy_snapshots: List[Any] = [None] * self.window
+        self.snapshot_counter: int = 0
+        self.learning_behavior_name: str = None
+        self.current_policy_snapshot = None
+        self.last_step = 0
+
+        # Chosen because it is the initial ELO in Chess
+        self.initial_elo: float = 1200.0
+        self.current_elo: float = self.initial_elo
+        self.policy_elos: List[float] = [self.initial_elo] * self.window
+        self.current_opponent: int = 0
+
+    @property
+    def get_step(self) -> int:
+        """
+        Returns the number of steps the trainer has performed
+        :return: the step count of the trainer
+        """
+        return self.trainer.get_step
+
+    def _write_summary(self, step: int) -> None:
+        """
+        Saves training statistics to Tensorboard.
+        """
+        opponents = np.array(self.policy_elos)
+        LOGGER.info(
+            " Learning brain {} ELO: {}\n"
+            "Mean Opponent ELO: {}"
+            " Std Opponent ELO: {}".format(
+                self.learning_behavior_name,
+                self.current_elo,
+                opponents.mean(),
+                opponents.std(),
+            )
+        )
+        self.stats_reporter.add_stat("ELO", self.current_elo)
+
+    def _process_trajectory(self, trajectory: Trajectory) -> None:
+        if (
+            trajectory.done_reached
+            and not trajectory.max_step_reached
+            and self.current_opponent > -1
+        ):
+            result = "win"
+            final_reward = trajectory.steps[-1].reward
```

Comment:
Do we ever want to factor in the magnitude of the reward and/or the opponent's reward? If not, it might be worth calling out in the docs that only the sign of the reward determines the winner.

Reply:
I called this out in the doc (still under construction).